### PR TITLE
fix: raise `SyntaxError` for unparenthesized assert and assign

### LINKED ISSUE #13070: TypeScript: unparenthesized `as` in assignment LHS
## Bug Report

**Current behavior**

- [REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=DYUwLgBARhC8EHIEBoIA84QN4QIYC4IBmCAXwG4AoSgCiwOjwGc8A7AT0wCYAGP0zGgCUVGLha4Og8kA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=typescript&prettier=false&targets=&version=7.13.13&externalPlugins=), [TypeScript Playground](https://www.typescriptlang.org/play?#code/DYUwLgBARhC8EHIEG4BQpIA84QN4QEMAuCAZggF81UAKXY6QgZ0IDsBPHAJgAY+KcmAJTIgA)

**Input Code**

```js
({a: b as any = 2000} = x);
b as any = x;
```

**Expected behavior**
Syntax error

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**

```js
{ plugins: ["typescript"] }
```

**Environment**

- Babel version(s): v7.13.10, v7.13.13

**Additional context**

If the `as`-expressions had parentheses around them, it'd be valid TypeScript syntax.

### PR DESCRIPTION
**Pull request by fedeci**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #13070  
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | Nope
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Since there is no way to know the previous node we check the `contextDescription` to have a generic understanding of what it was. However checking that string may lead to errors in the future if the value changes. We may create an enum to store those like is done for `Errors`.

### CODE REVIEW

File: packages/babel-parser/test/fixtures/typescript/cast/unparenthesized-type-assertion-and-assign/input.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+<string>foo = '100';
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/test/fixtures/typescript/cast/unparenthesized-type-assertion-and-assign/input.ts:1]: Can you also add a test for `(<string> foo) = "100"`? (unless we already have it)

Comment (fedeci) [reply packages/babel-parser/test/fixtures/typescript/cast/unparenthesized-type-assertion-and-assign/input.ts:1]: We have it https://github.com/babel/babel/blob/7bc72bb451e71d7d8964efdf4438d5a907aae655/packages/babel-parser/test/fixtures/typescript/cast/parenthesized-assert-and-assign/input.ts#L1

Comment (nicolo-ribaudo) [reply packages/babel-parser/test/fixtures/typescript/cast/unparenthesized-type-assertion-and-assign/input.ts:1]: I meant with the `TSTypeAssertion` syntax, not with `TSAsExpression`

Comment (fedeci) [reply packages/babel-parser/test/fixtures/typescript/cast/unparenthesized-type-assertion-and-assign/input.ts:1]: Ops wrong file🤦 https://github.com/babel/babel/blob/main/packages/babel-parser/test/fixtures/typescript/cast/parenthesized-type-assertion-and-assign/input.ts

Comment (nicolo-ribaudo) [reply packages/babel-parser/test/fixtures/typescript/cast/unparenthesized-type-assertion-and-assign/input.ts:1]: Oh ok 👍

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by existentialism — APPROVED** [review]

**Pull request merged**